presto-main/src/main/java/com/facebook/presto/sql/analyzer/ScopeReferenceExtractor.java

```diff
@@ -41,53 +38,25 @@ public static boolean hasReferencesToScope(Node node, Analysis analysis, Scope s
 
     public static List<Expression> getReferencesToScope(Node node, Analysis analysis, Scope scope)
     {
-        ImmutableList.Builder<Expression> builder = ImmutableList.builder();
-        new Visitor(analysis, scope).process(node, builder);
-        return builder.build();
-    }
-
-    private static class Visitor
-            extends DefaultExpressionTraversalVisitor<Void, ImmutableList.Builder<Expression>>
-    {
-        private final Analysis analysis;
-        private final Scope scope;
-
-        private Visitor(Analysis analysis, Scope scope)
-        {
-            this.analysis = requireNonNull(analysis, "analysis is null");
-            this.scope = requireNonNull(scope, "scope is null");
-        }
-
-        @Override
-        protected Void visitIdentifier(Identifier node, ImmutableList.Builder<Expression> context)
-        {
-            if (!analysis.getLambdaArgumentReferences().containsKey(node) && isReferenceToScope(node, QualifiedName.of(node.getName()))) {
-                context.add(node);
-            }
-            return null;
-        }
+        Map<Expression, FieldId> columnReferences = analysis.getColumnReferenceFields();
 
-        @Override
-        protected Void visitLambdaExpression(LambdaExpression node, ImmutableList.Builder<Expression> context)
-        {
-            return process(node.getBody(), context);
-        }
+        List<Expression> referencesToScope = AstUtils.preOrder(node)
+                .filter(columnReferences::containsKey)
+                .map(Expression.class::cast)
+                .filter(expression -> isReferenceToScope(expression, scope, columnReferences))
+                .collect(toImmutableList());
 
-        @Override
-        protected Void visitDereferenceExpression(DereferenceExpression node, ImmutableList.Builder<Expression> context)
-        {
-            if (analysis.getColumnReferences().contains(node) && isReferenceToScope(node, DereferenceExpression.getQualifiedName(node))) {
-                context.add(node);
-                return null;
-            }
+        return referencesToScope;
+    }
 
-            return super.visitDereferenceExpression(node, context);
-        }
+    private static boolean isReferenceToScope(Expression node, Scope scope, Map<Expression, FieldId> columnReferences)
+    {
+        FieldId fieldId = requireNonNull(columnReferences.get(node), () -> "No FieldId for " + node);
+        return isFieldFromScope(fieldId, scope);
+    }
 
-        private boolean isReferenceToScope(Expression node, QualifiedName qualifiedName)
-        {
-            Optional<ResolvedField> resolvedField = scope.tryResolveField(node, qualifiedName);
-            return resolvedField.isPresent() && resolvedField.get().getScope().equals(scope);
-        }
+    public static boolean isFieldFromScope(FieldId fieldId, Scope scope)
```

Comment:
Do we have such a rule in general? Here, in this class, the methods are from top-to-bottom in order of dependency. Helps reading, but sure I can move.